core/connection.py

```
@@ -47,8 +47,16 @@ def _path(self, client_id: str) -> str:
-            with open(self._path(client_id), "w") as f:
+            # Atomic write so a crash mid-write, or a concurrent startup scan,
+            # never reads a half-written file (which the loader would then drop
+            # as "unreadable", losing a live bot's state).
+            path = self._path(client_id)
+            tmp = f"{path}.{os.getpid()}.tmp"
+            with open(tmp, "w") as f:
                 json.dump(list(details), f)
+                f.flush()
+                os.fsync(f.fileno())
+            os.replace(tmp, path)
         except Exception as e:
```

Ruff (0.15.21)

[warning] Do not catch blind exception: `Exception`

(BLE001)